optimization/sggp/sggp-algorithm-with-test.metta

```diff
@@ -0,0 +1,597 @@
+                            (generateTrees $lhs $grammarSpace $stocSpace $explorationCoeff (- $depth 1))
+                            ;; (This is calculated probability $calculatedProbability)
+                            ;; (This is random value (random))
+                        )
+                        
+                    )
+            )
+
+            )
+        )
+    ;
+        
+    )
+)
+;; (= (generate $parentSymbol $initSymbol $grammarSpace $stocSpace $depth)
+;;     (if (== (findSuitableProdRules $initSymbol $grammarSpace) ())
+;;         (findAndReplace $parentSymbol $initSymbol)
+;;     )
+;; )
+;; The following function is an adapter that integrates the scoring 
+;; to the dataset
+(= (fitnessWithdata $data)
+    (fitness $data
+             ((True (x1 True) (x2 False)) 
+             (False (x1 False) (x2 False)) 
+           (True (x1 True) (x2 True))))
+
+)
+;; This algorithm implements the main loop of the sggp algorithm
+;; it works by generating trees from the stochastic grammar values
+;; it filters the trees as good trees and bad trees
+;; it reinforces the stochastic grammar based on the two classified syntax trees
+;; The threshold is specified to the main loop which is given to the generic function 
+;; called filterlist.
+(= (sggpHelper $initSymbol $learnRate $mutationRate $thresholdProb $maxRecur $explorationCoeff $maxDepth $scoreFunc)
+    (let () (println! ($maxRecur))
+    (if (> $maxRecur 0)
+        (let* (
+            ($generatedInstances (collapse (unique (generateTrees $initSymbol &grammarSpace &stochasticGrammarSpace $explorationCoeff $maxDepth))))
+            ($_ (println! (generatedInstances ==> $generatedInstances)))
+            ;;($collapsedValues (collapse $generatedInstances))
+            ($goodTrees (collapse (filter' >= $scoreFunc (superpose $generatedInstances) $thresholdProb)))
+            ($badTrees (collapse (filter' < $scoreFunc (superpose $generatedInstances) $thresholdProb)))
+            ($_ (println! (goodTrees $goodTrees badTrees $badTrees)))
+            ($lenGood (len $goodTrees))
+            ($lenBad (len $badTrees))
+            ($reinforced (collapse (reinforceTreeRulesIterator  $goodTrees $learnRate $lenGood &grammarSpace &stochasticGrammarSpace)))
+            ($punished (collapse (punishTreeRulesIterator  $badTrees $learnRate $lenBad &grammarSpace &stochasticGrammarSpace)))
+        )
+            (let () (println! (goodTrees ==> $goodTrees))
+                (sggpHelper $initSymbol $learnRate $mutationRate $thresholdProb (- $maxRecur 1) $explorationCoeff $maxDepth $scoreFunc)
+            )
+            
+        )
+        ;; (generateInstances junctor &grammarSpace &stochasticGrammarSpace $maxDepth)
+        (collapse (unique (generateTrees $initSymbol &grammarSpace &stochasticGrammarSpace $explorationCoeff $maxDepth)))
+    )
+)
+)
+;;Helper for a reinforceGrammarBased on Tree
+(=(reinforceTreeHelper $distributedProdRules $learnRate $cardinality $stocSpace)
```

Comment:
Best if you can move this function to just below the `reinforceGrammarbasedOnTree` function, for better readability.